Dashboard v2: Account Profile - Update email

pandah3 opened the pull request:
Linear: https://linear.app/a8c/issue/DOTDASH-514/implement-profile-update-email-verification-flow

## Proposed Changes

Implementing the email verification flow to update your user profile email address.

**Pending state:**
<img width="773" height="697" alt="Screenshot 2025-09-29 at 4 10 04 AM" src="https://github.com/user-attachments/assets/7c0d03b2-025d-41cc-b63c-3056bacb31b1" />

**Success state - existing user changing emails**

<img width="743" height="609" alt="Screenshot 2025-10-01 at 1 38 50 AM" src="https://github.com/user-attachments/assets/10877371-7d42-42e7-aae7-0325d52d9bd1" />


**Success state - new user (no link, updated messaging)**
<img width="746" height="582" alt="Screenshot 2025-10-01 at 1 38 58 AM" src="https://github.com/user-attachments/assets/aeceaa6f-2b90-475e-b34a-a31f66248c56" />


**Error state**
<img width="1299" height="776" alt="Screenshot 2025-10-01 at 1 39 33 AM" src="https://github.com/user-attachments/assets/7e77b0bc-9f67-4329-971d-c12e44064e17" />


There are no design mockups for the update email flow so I based the changes off of v1, as well as the Security - Account Recovery email section for consistency across the dashboard. I used the same notice for the pending email verification state + the success notice outlined in Figma for the Account Recovery pages (QOBjujZah0UlGDDdLKZhzi-fi-8995-151325).

**Note about the v1/v2 redirect:**
When you click on the verification link in the email you receive, it takes you to v1. I played around with various ways to land on the v2 profile page instead, but nothing worked super reliably (or at all). Some backend changes may be in order.

_Edit: Redirect working now._

For now, I mirrored the implementation to match the Hosting Dashboard opt-in logic to redirect the user to v2 if they opted into the HD. I had my user opt-in from the Calypso side, but I still couldn't get the redirect to v2 to work. Not sure if there's something I'm missing here.

## Why are these changes being made?

* So users can change their email address!

## Testing Instructions

* `yarn start`
* Visit http://calypso.localhost:3000/v2/me/profile
* Personal details -> Email address field: Update your email address. Make sure validations for the input works.
* Click save -> ensure you see the "Verify email" notice
* Try canceling the request + resending the email and make sure both work
* Make sure you get a verification email in your inbox
* When you click the verification email, the eventual expected behavior is: verified in v1, then redirected to v2. But the verification link in the email will take you to _production_ v1. To test the redirect logic locally, navigate to http://calypso.localhost:3000/me/account?new_email_result=1 .
* You should be redirected to /v2/me/profile and see the green success notice with the domains link
* You can play around with testing the redirect behavior or just the success/error messages. `/me/account?<param=value>` will redirect you to v2. `/v2/me/profile?<param=value>` will let you see the different messages without the redirect:
  * Existing user verifying new email to change to:
    * Success: `?new_email_result=1` . Should see green success notice with the domains link
    * Error: `?new_email_result=0` . Should see error snackbar notice
  * New user verifying email for the first time:
    * Success: `?verified=1` . Should see green success notice with no link and "Email verified" instead of "updated"
    * Error: `?verified=0` . Should see the same error snackbar notice as the other case

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

pandah3 committed:
Setup backend for email verification:
- Create API mutations in /api-core
- Update the UserSettings types to include the email verification fields
- Add the mutations to the /api-queries package

pandah3 committed:
Merge branch 'trunk' into add/user-email-verification

pandah3 committed:
Add email verification notice component for verification pending state:
- Show info notice when email change is pending
- Add resend verification email functionality
- Handle email verification URL parameters for verification success/error

pandah3 committed:
Add <EmailSection> component to handle the user email input and updating logic:
- Add pending email state: disable field, show cancel option, etc.
- Add logic for canceling pending verification
- Input validation + inline helper messages
- Render email field in profile section

pandah3 committed:
Invalidate settings on cancel to make the input field available right away. Fix lint and type errors

pandah3 committed:
Add email verification redirect support for v2 HD users: If user opted into HD, redirect from the v1 email verification page (me/account) to v2 profile page

pandah3 committed:
Change success state notification to a more persistent notice since it includes a link for users to click.

pandah3 committed:
Add error handling for email validator + null check for pendingEmail in resend function. Fix lint errors

pandah3 committed:
Add tests

pandah3 committed:
Fix failing tests: Add missing email verification fields to mock user settings, mock new email components to avoid complexity, etc.

pandah3 committed:
Merge branch 'trunk' into add/user-email-verification

andrea-sdl commented on client/components/email-verification/index.js:
```diff
@@ -12,6 +13,20 @@ export default function emailVerification( context, next ) {
+		const state = context.store.getState();
+		const hostingDashboardPreference = getPreference( state, 'hosting-dashboard-opt-in' );
+
+		if ( hostingDashboardPreference === 'opt-in' ) {
```
There's a dedicated utility for this that I think you should be able to use https://github.com/Automattic/wp-calypso/blob/788b124c58a45630a383f58de2a4b1dc07982b37/client/state/sites/selectors/has-hosting-dashboard-opt-in.ts#L6

pandah3 replied:
Updated in [302dded](https://github.com/Automattic/wp-calypso/pull/106103/commits/302dded34dd9cb6c774ab8278932fea347001fdc), thank you!!

andrea-sdl commented on client/components/email-verification/index.js:
```diff
@@ -12,6 +13,20 @@ export default function emailVerification( context, next ) {
+		const hostingDashboardPreference = getPreference( state, 'hosting-dashboard-opt-in' );
+
+		if ( hostingDashboardPreference === 'opt-in' ) {
+			window.location.href = `/v2/me/profile?new_email_result=${ context.query.new_email_result }`;
```
I'm not sure this is working. I tried by testing it on `/me/account?verified=1&new_email_result=1` and I wasn't redirected. I know you mentioned some backend changes might be required but I wonder if there's something else at play here.

pandah3 replied:
Yeah I wasn't having any success either but I think I figured it out! My state had the correct HD opt-in preference, but it wasn't picking it up for some reason (I had the correct `hostingDashboardPreference.value` when I was testing). Now that I wait for it to load first, the redirect works for me: [302dded](https://github.com/Automattic/wp-calypso/pull/106103/commits/302dded34dd9cb6c774ab8278932fea347001fdc)

For the backend changes, I meant that it'd be good to not have to play with this redirect stuff and just send the user to the correct route based on the v1/v2 version. At least the redirect is working for now!!

Side note: `/me/account?new_email_result=1` should work. I don't know if we have both params together.

pandah3 replied:
This was refactored to accommodate for more redirect cases. We can now do these four and they should all redirect appropriately! See https://github.com/Automattic/wp-calypso/pull/106103#issuecomment-3355446231 for more details.

- me/account?new_email_result=1
- /me/account?new_email_result=0
- /me/account?verified=1
- /me/account?verified=0

andrea-sdl commented on packages/api-core/src/me-settings/mutators.ts:
```diff
@@ -32,9 +32,18 @@ export async function updateUserSettings(
 }
+
+export async function cancelPendingEmailChange(): Promise< Partial< UserSettings > > {
+	return await wpcom.req.post( '/me/settings', { user_email_change_pending: false } );
```
nitpick, but I think we could use the `updateUserSettings` function here

pandah3 replied:
I called the endpoint intentionally because I don't think we need to go through all the other fields in `userSettings` just to update the 1-2 fields. Plus, it's an actual "action" mutation to cancel the email rather than just updating the settings in general, so I thought it was just best to call the endpoint directly.

andrea-sdl replied:
Yup, fine for me. I don't have a strong opinion about this ;)

fushar replied:
> I don't think we need to go through all the other fields in userSettings just to update the 1-2 fields.

Could you clarify this? I think we can just simply do something like this in the mutation?

```ts
mutationFn: () => updateUserSettings( { user_email_change_pending: false } ),
```

pandah3 replied:
Ah okay, I misunderstood! I still think the semantics of "canceling" vs "updating" is good to differentiate, but I updated it in [9908200](https://github.com/Automattic/wp-calypso/pull/106103/commits/990820088e056b918dfeba7a9780d05b3b815411)!

andrea-sdl commented on client/components/email-verification/index.js:
```diff
@@ -12,6 +13,20 @@ export default function emailVerification( context, next ) {
+		const state = context.store.getState();
+		const hostingDashboardPreference = getPreference( state, 'hosting-dashboard-opt-in' );
+
+		if ( hostingDashboardPreference === 'opt-in' ) {
```
```suggestion
		if ( hostingDashboardPreference.value === 'opt-in' ) {
```
Here's what's missing to make it work ;)

pandah3 replied:
Oops, looks like this fell off as I was testing things 🙈 Used the `hasHostingDashboardOptIn` helper in [302dded](https://github.com/Automattic/wp-calypso/pull/106103/commits/302dded34dd9cb6c774ab8278932fea347001fdc)!

andrea-sdl commented on client/dashboard/me/profile-personal-details/test/email-section.test.tsx:
```diff
@@ -0,0 +1,99 @@
+	useSuspenseQuery: jest.fn(),
+	useMutation: jest.fn(),
+} ) );
+
```
@pandah3 do we need to mock everything in this scenario? In the past I've had some good success with mocking just the API result (example below)
https://github.com/Automattic/wp-calypso/blob/d20fbdec22d14ce1e595787ba857879a023e043e/client/dashboard/me/preferences-language/test/index.test.tsx#L13

pandah3 replied:
Good point! I ended up adding this because some tests were erroring out but I overmocked. The `nock` worked beautifully, thank you!! Updated in [8cbb62b](https://github.com/Automattic/wp-calypso/pull/106103/commits/8cbb62b321cc73c73895907a5f2541511a86c231).

andrea-sdl commented on client/dashboard/me/profile-personal-details/test/index.test.tsx:
```diff
@@ -22,6 +22,31 @@ jest.mock( '../update-username/username-validation-utils', () => ( {
+		return null;
+	};
+} );
+
```
```suggestion
```
I think we can skip this and just keep the `cancelPendingEmailChangeMutation/resendEmailVerificationMutation` mock as it worked fine IMO.

pandah3 replied:
Yeah, this was needed because of the `react-query` mock. Removed now that we're not mocking that anymore!! [8bcc572](https://github.com/Automattic/wp-calypso/pull/106103/commits/8bcc572e5f88b4609bb7f7b2ca455352ccc9ef8b)

andrea-sdl commented on client/dashboard/me/profile-personal-details/update-email/test/email-verification-banner.test.tsx:
```diff
@@ -0,0 +1,131 @@
+	...jest.requireActual( '@tanstack/react-query' ),
+	useSuspenseQuery: jest.fn(),
+	useMutation: jest.fn(),
+} ) );
```
same question as here https://github.com/Automattic/wp-calypso/pull/106103/files#r2391228232 , not sure we need this much mocking.

pandah3 replied:
Updated in [8cbb62b](https://github.com/Automattic/wp-calypso/pull/106103/commits/8cbb62b321cc73c73895907a5f2541511a86c231)!

andrea-sdl commented on client/dashboard/me/profile-personal-details/update-email/email-verification-banner.tsx:
```diff
@@ -0,0 +1,129 @@
+				onClose={ () => setShowSuccessNotice( false ) }
+				actions={
+					<Button variant="link" href="/v2/domains/">
+						{ __( 'Update domain contacts' ) }
+					</Button>
+				}
```
```suggestion
				actions={ <Link to="/domains/">{ __( 'Update domain contacts' ) }</Link> }
```

I noticed that clicking the domains link will trigger a full page refresh. Using `import { Link } from '@tanstack/react-router';` should fix that

pandah3 replied:
Ahh and using `Button` here isn't accessible either 🙈 Updated to use `<Link>`, thank you!! [8bcc572](https://github.com/Automattic/wp-calypso/pull/106103/commits/8bcc572e5f88b4609bb7f7b2ca455352ccc9ef8b)

andrea-sdl reviewed (commented):
Left some comments, overall it's working great (besides the redirect issue) 🎉 

I also like the fact that, compared to the previous experience, we're signaling that the link has expired

<img width="1104" height="240" alt="Image" src="https://github.com/user-attachments/assets/fa6d7422-1a14-4006-b6ba-ac33712c68c0" />

pandah3 committed:
Fix email verification redirect timing: wait for preferences to load before checking for HD opt-in status.

pandah3 committed:
Update tests to use nock to mock API responses and properly wait for the render so we're testing the actual API interactions

pandah3 committed:
Remove unnecessary test and use React router link to avoid a full page refresh

pandah3 committed:
Merge branch 'trunk' into add/user-email-verification

pandah3 committed:
Redirect users to v2 if they opted into the HD even if the email verification fails

pandah3 committed:
Refactor email verification logic to handle v1 and v2 logic more cleanly:
- Split out v1 and v2 logic verification logic. v2 redirects from v1 and handles success and error cases. v1 handles success cases only.
- Redirect to v2 for new users who are verifying their email for the first time as well
- Conditionally show different success messages after verification depending on the user journey. Don't show the domains link if it's a new user.

pandah3 committed:
Fix email verification UI logic:
- Update the useEffect to store the verification type before cleaning up the URL
- Initialize state from URL params outside of useEffect so there's no delay
- Fix UI flashing issue by removing useSuspenseQuery to prevent double data fetching and rely on userData prop from parent instead
- Clean up repetitive code into helper functions
- Correctly show messages based on the email verification type and status

pandah3 committed:
Merge branch 'trunk' into add/user-email-verification

pandah3 commented:
@andrea-sdl Thank you for the review!! I addressed all the feedback + added some more redirect cases because I noticed there's more than one type of email validation:

- Existing user changing emails and verifying the new email - success + error case
- New user verifying email for the first time - success + error case

I added the additional 3 cases besides the one existing user success redirect that we had.
Updated the testing steps in the PR description and added more screenshots to show the different states.

fushar commented on client/dashboard/me/profile-personal-details/email-section.tsx:
```diff
@@ -0,0 +1,159 @@
+		onSuccess: () => {
+			onChange( userData.user_email || '' );
+			// Invalidate user settings to refetch updated data on cancel
+			queryClient.invalidateQueries( userSettingsQuery() );
```
Please move invalidation to the query layer, in the `cancelPendingEmailChangeMutation()`'s onSuccess itself. The idea is that components shouldn't care about invalidations; it should be handled by the queries.

pandah3 replied:
Got it! Updated in [1f5214f](https://github.com/Automattic/wp-calypso/pull/106103/commits/1f5214f07efb55ef3e2b76b49c5c805e8cbf8a99), thank you!

pandah3 replied:
Had to update this in a different way bc the cache invalidation in the mutation wasn't working as expected: https://github.com/Automattic/wp-calypso/pull/106103#issuecomment-3357171127

pandah3 committed:
Move cache invalidation to the query layer so it's handled by the query and not the component

andrea-sdl commented on client/dashboard/me/profile-personal-details/update-email/email-verification-banner.tsx:
```diff
@@ -0,0 +1,169 @@
+	}
+	const params = new URLSearchParams( window.location.search );
+	return {
+		newEmailResult: params.get( 'new_email_result' ),
+		verified: params.get( 'verified' ),
```
I was thinking that through the code we use lots of ` === '1'` and `=== '0'`.
Not a requirement for the PR but a thought: Maybe this function could return 4 outputs

```suggestion
		isEmailChangeComplete: params.get( 'new_email_result' ) === '1',
		isEmailVerificationComplete: params.get( 'verified' ) === '1',
        emailChangeFailed: params.get( 'new_email_result' ) === '0',
        emailVerificationFailed: params.get( 'verified' ) === '0',
```

And we use them right where we need them. This way the logic is gated in one single entry point and we don't need to think about what the `1|0` value means.

pandah3 replied:
Implemented in [a3f1750](https://github.com/Automattic/wp-calypso/pull/106103/commits/a3f1750231d910b78a2261c28a11ad386304f81a)!

andrea-sdl commented on client/dashboard/me/profile-personal-details/update-email/email-verification-banner.tsx:
```diff
@@ -0,0 +1,169 @@
+	const { createErrorNotice } = useDispatch( noticesStore );
+	const [ showResendButton, setShowResendButton ] = useState( true );
+	const [ showSuccessNotice, setShowSuccessNotice ] = useState( () => {
+		const { newEmailResult, verified } = getEmailVerificationParams();
```
nitpick: I think we can extract this line outside of the clojure function and use those two fields everywhere they are need with a single function call.
Example
```suggestion
	const { newEmailResult, verified } = getEmailVerificationParams();
	const [ showSuccessNotice, setShowSuccessNotice ] = useState( () => {
```

pandah3 replied:
Extracted in [fe5690f](https://github.com/Automattic/wp-calypso/pull/106103/commits/fe5690fa907ae1c6f29e21417fd31a677bd287f3).

andrea-sdl reviewed (commented):
The flow works as expected, but I've hit some issues (see p1759326488544079-slack-CKUJZT08Y ) with the "Cancel the pending email change flow)

Specifically, the email field stays locked in (disabled status), with no ways to edit it again. 

<img width="2982" height="2384" alt="Image" src="https://github.com/user-attachments/assets/54d27418-4f71-4c03-b9e8-52b0c1e5c31b" />
A page refresh helps, although for a split second we see the previous config (probably because TanStack is picking it up from the cache).

pandah3 committed:
Fix email field staying disabled after canceling pending email change:
- Preserve original mutation's onSuccess handler when extending in component
- Ensure cache updates from query layer are not overridden by component logic
- Pass userData from parent to avoid stale closures

pandah3 committed:
Refactor email verification params for clarity and easier readability:
- Replace raw '1'/'0' string comparisons with descriptive boolean flags
- Centralize URL parameter interpretation logic, but keep v1 and v2 implementations separate to maintain clean architecture boundaries

pandah3 committed:
Extract verification params once to avoid multiple URL parsing calls

pandah3 committed:
Use updateUserSettings directly in cancelPendingEmailChangeMutation instead

pandah3 committed:
Merge branch 'trunk' into add/user-email-verification

pandah3 commented:
@andrea-sdl Fixed the email field staying disabled after canceling pending email change in [5a9b13f](https://github.com/Automattic/wp-calypso/pull/106103/commits/5a9b13fda7c2da53baa478e478bbb18f9f43a0a5)!

The cache invalidation wasn't working bc the cache updates from query layer were overridden by component logic so we needed to preserve the original mutation's `onSuccess` handler.

andrea-sdl commented on client/dashboard/me/profile-personal-details/update-email/email-verification-banner.tsx:
```diff
@@ -0,0 +1,185 @@
+
+// Get email verification params from URL
+function getEmailVerificationParams() {
+	if ( typeof window === 'undefined' ) {
```
@pandah3 I'm mostly curious: but is this for the tests or have you faced the error of window being undefined?

pandah3 replied:
Just defensive practice I guess 😅 Probably not super necessary for HD, but I think it's harmless to keep. Will leave in there for now, but I can always remove it!

andrea-sdl commented on client/dashboard/me/profile-personal-details/update-email/email-verification-banner.tsx:
```diff
@@ -0,0 +1,185 @@
+	const [ showResendButton, setShowResendButton ] = useState( true );
+
+	// Extract verification params once to avoid multiple URL parsing calls
+	const verificationParams = getEmailVerificationParams();
```
```suggestion
	const {isEmailVerificationComplete,isEmailChangeComplete} = getEmailVerificationParams();
```

Something like this would probably help with the `verificationParams` repetition below (not a requirement, but I thought it was a little verbose).

pandah3 replied:
Ahh yes, gotta love destructuring! Thanks for the suggestion! Applied in [2a75a25](https://github.com/Automattic/wp-calypso/pull/106103/commits/2a75a2594fc9a0714c3590f0f16c360a85fc5f8e)!

andrea-sdl reviewed (approved):
Looks good! I tested the canceling too and now it perfectly reloads the data. Great job on iterating on the feedback. I left a smaller comment, not a blocker ;)

andrea-sdl commented on packages/api-queries/src/me-settings.ts:
```diff
@@ -22,3 +26,33 @@ export const userSettingsMutation = () =>
+
+export const resendEmailVerificationMutation = () =>
+	mutationOptions( {
+		mutationFn: resendEmailVerification,
```
@pandah3 not sure if you also want to change this to call the `updateUserSettings` like you did on the cancel part. raising this for awareness.

pandah3 replied:
Updated in [f0fa4e5](https://github.com/Automattic/wp-calypso/pull/106103/commits/f0fa4e500bb4df78b0b9eef389a4ccca7e246ec3)!

pandah3 committed:
Destructure getEmailVerificationParams() for easier readability

pandah3 committed:
Use updateUserSettings directly in resendEmailVerificationMutation instead of calling the endpoint

pandah3 committed:
Merge branch 'trunk' into add/user-email-verification

pandah3 commented:
Applied final additional feedback improvements, woohoo!
